src/exo/cursors.py

```diff
@@ -0,0 +1,591 @@
+        return p
+
+    # ------------------------------------------------------------------------ #
+    # Navigation (abstract)
+    # ------------------------------------------------------------------------ #
+
+    @abstractmethod
+    def parent(self) -> Node:
+        """Get the node containing the current cursor"""
+
+    @abstractmethod
+    def before(self, dist=1) -> Cursor:
+        """For gaps, get the node before the gap. Otherwise, get the gap before
+        the node or block"""
+
+    @abstractmethod
+    def after(self, dist=1) -> Cursor:
+        """For gaps, get the node after the gap. Otherwise, get the gap after
+        the node or block"""
+
+    @abstractmethod
+    def prev(self, dist=1) -> Cursor:
+        """Get the previous node/gap in the block. Undefined for blocks."""
+
+    @abstractmethod
+    def next(self, dist=1) -> Cursor:
+        """Get the next node/gap in the block. Undefined for blocks."""
+
+    # ------------------------------------------------------------------------ #
+    # Protected path / mutation helpers
+    # ------------------------------------------------------------------------ #
+
+    def _translate(self, ty, path, dist):
+        if not path:
+            raise InvalidCursorError('cannot move root cursor')
+        attr, i = path[-1]
+        if i is None:
+            raise InvalidCursorError('cursor is not inside block')
+        return ty(self._proc, path[:-1] + [(attr, i + dist)])
+
+    @staticmethod
+    def _walk_path(n, path):
+        for attr, idx in path:
+            n = getattr(n, attr)
+            if idx is not None:
+                n = n[idx]
+        return n
+
+    def _rewrite_node(self, fn):
+        """
+        Applies `fn` to the AST node containing the cursor. The callback gets
+        the following arguments:
+        1. The parent node
+        2. The relevant list of children
+        3. The last step in the path (the attribute and index info)
+        The callback is expected to return a single, updated node to be placed
+        in the new tree.
+        """
+        assert isinstance(self, (Node, Block, Gap))
+
+        def impl(node, path):
+            if len(path) == 1:
+                return fn(node)
+
+            (attr, i), path = path[0], path[1:]
+            children = getattr(node, attr)
+
+            if i is None:
+                return node.update(**{attr: impl(children, path)})
+
+            return node.update(**{
+                attr: children[:i] + [impl(children[i], path)] + children[i + 1:]
+            })
+
+        return impl(self.proc().INTERNAL_proc(), self._path)
+
+    def _make_forward(self, new_proc, fwd_node, fwd_gap, fwd_sel):
+        orig_proc = self._proc
+        depth = len(self._path)
+        attr = self._path[depth - 1][0]
+
+        def forward(cursor: Cursor) -> Cursor:
+            if cursor._proc != orig_proc:
+                raise InvalidCursorError('cannot forward unknown procs')
```

Comment:
Should this be an `assert` error (i.e. internal) or can a user trigger this?

Reply:
A user can trigger this by trying to use a cursor derived from one proc in a scheduling directive applied to a different proc.

Reply:
That should be caught and signaled by the provenance tracking code then, no?

Reply:
i.e. some other (internal) code is going to have to find the right sequence of forwarding functions to apply anyway, and if it applies the wrong forwarding function then that would be an error in that other internal code, not a user error.

Reply:
As we build it out, checks in that part of the system might supersede checks here. Then this could become a true assertion.